arbor/communication/distributed_for_each.hpp

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "distributed_context.hpp"
+#include "util/range.hpp"
+
+namespace arb {
+
+namespace impl {
+template <class FUNC, typename... T, std::size_t... Is>
+void for_each_in_tuple(FUNC&& func, std::tuple<T...>& t, std::index_sequence<Is...>) {
+    (func(Is, std::get<Is>(t)), ...);
+}
+
+template <class FUNC, typename... T>
+void for_each_in_tuple(FUNC&& func, std::tuple<T...>& t) {
+    for_each_in_tuple(func, t, std::index_sequence_for<T...>());
+}
+
+template <class FUNC, typename... T1, typename... T2, std::size_t... Is>
+void for_each_in_tuple_pair(FUNC&& func,
+    std::tuple<T1...>& t1,
+    std::tuple<T2...>& t2,
+    std::index_sequence<Is...>) {
+    (func(Is, std::get<Is>(t1), std::get<Is>(t2)), ...);
+}
+
+template <class FUNC, typename... T1, typename... T2>
+void for_each_in_tuple_pair(FUNC&& func, std::tuple<T1...>& t1, std::tuple<T2...>& t2) {
+    for_each_in_tuple_pair(func, t1, t2, std::index_sequence_for<T1...>());
+}
+
+}  // namespace impl
+
+
+
+/*
+ * Iterate through multiple ranges from each distributed rank. Only usable with trvially copyable value types.
+ * The provided function objects is expected to be callable with the following signature:
+ * (const util::range<util::range<ARGS>::value_type*>&...) -> void
+ * Given 'n' distributed ranks, the function will be called 'n' times with data from each rank.
+ * There is no guaranteed order.
+ */
+template <typename FUNC, typename... ARGS>
```

Comment:
Suggested documentation from what I read in the code:

Collective operation, calling `func` on `args` supplied by each rank exactly once. The order of calls is unspecified.
Requires
- `Item = args::value_type` to be identical across all ranks
- `Item` is `trivially_copyable`
- `func` to be a callable type with signature `void func(util::range<const Item> range)`
- `func` _must not_ modify contents of `range`
- All ranks in `distributed` _must_ call this collectively.

`for_each` is the right name, yes? `func` may be an object with some internal accumulator, e.g. 
```cxx
struct Counter {
  std::size_type n = 0;
  template<typename T> void operator()(const range<T>& rg) { n += rg.size(); }
};
```
and will do what one would expect?

Reply:
I've updated the documentation with your suggestion.
The example should give the expected total size, yes.